Add Language support for OpenAI title and excerpt generation

phpbits opened the pull request:
### Description of the Change
This PR adds Language Settings that applies to both Title and Excerpt generation.
<img width="1479" alt="Screen Shot 2023-07-14 at 11 40 23 PM" src="https://github.com/10up/classifai/assets/3365507/5c50c85f-4962-4a59-bf35-358e7d4f9480">


<!-- Enter any applicable Issue number(s) here that will be closed/resolved by this PR. -->
Relates to [#150](https://github.com/10up/classifai/issues/150)

### How to test the Change
<!-- Please provide steps on how to test or validate that the change in this PR works as described. -->
Admin should get results in the preferred language.
Switch the following when testing user cases -
- The classic editor/ The Gutenberg editor
- Use different language

#### Test
 - [ ] Visit the edit post and click on the 'Generate Titles' button 
 - [ ] Visit the edit post and click on the 'Generate/ Re-Generate Excerpt' button 
 


### Changelog Entry
<!--
Please include a summary for this PR, noting whether this is something being Added / Changed / Deprecated / Removed / Fixed / or Security related.  You can replace the sample entries after this comment block with the single changelog entry line for this PR. -->
> Added - Add Language support for OpenAI title and excerpt generation

### Credits
<!-- Please list any and all contributors on this PR so that they can be added to this projects CREDITS.md file. -->
Props @phpbits @ravinderk


### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you are unsure about any of these, please ask for clarification.  We are here to help! -->
- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my change.
- [ ] All new and existing tests pass.

phpbits committed:
Add Language support for OpenAI title and excerpt generation

phpbits committed:
Formatting fixes

phpbits committed:
Formatting fixes

phpbits commented:
@dkotter @jeffpaul This is now ready for review. Thanks!

dkotter commented:
@phpbits Thanks for this! Overall this looks okay to me and matches what I've seen other solutions do, where they update the prompt they send to OpenAI, asking ChatGPT to respond in a certain language. I'm wondering though if we can simplify this and remove the need for a new setting? So instead of pulling the language from a setting, we pull the site language and use that in the prompt?

jeffpaul commented:
Alternatively, defaulting the setting to the site language that way we don't require any other initial config from a site owner?

phpbits committed:
Update selected language to site language by default

phpbits committed:
Merge branch 'develop' into feature/language-support

phpbits commented:
Thanks, @dkotter and @jeffpaul. I used the assigned language as the default value for the new settings. I've also fixed the merge conflict.

ravinderk reviewed (changes requested):
@phpbits I text this with the Hindi language:
- [x] Excerpt generate in Hindi
- [ ] Title generate in Hindi

I am unsure whether it is an expected behavior because I was expecting the Title also in Hindi. Do you have thought on this?

sksaju committed:
added filter support on azure text to speech ssml tag

ravinderk commented:
@phpbits I can see a merge conflict on this pull request. Can you review them?

jeffpaul commented:
@ravinderk as you're able during your OSS time, could you please look to help wrap up any final tweaks needed here as @phpbits is unavailable for a bit?  Thanks!

ravinderk committed:
Merge branch 'develop' into feature/language-support

ravinderk committed:
feat: use language english name in prompt

ravinderk commented:
@dkotter I found that we are using a language code in the prompt, which is not giving the expected results in titles and excerpts. I used the English name of the language in the prompt, and each result was as expected.

I used this approach for the title. Let me know if you have any questions.

Testing Branch: https://github.com/10up/classifai/tree/alternative/feature/language-support
Commit: https://github.com/10up/classifai/commit/bb6d7eebc234d943e679f399f7d147044a22a42a

cc: @jeffpaul

dkotter commented:
@ravinderk 

> I used this approach for the title. Let me know if you have any questions.

This sounds like the correct approach to me, using the full language name instead of the language code.

ravinderk committed:
refactor: prevent php notices

We should check for language code in available translation only if language code is not 'en'

ravinderk committed:
refactor: create function to get language for promt

ravinderk committed:
feat: use helper function to get language for content excerpt prompt

ravinderk committed:
fix: set default language value to empty string

ravinderk committed:
doc: explain default value

ravinderk commented on includes/Classifai/Admin/SavePostHandler.php:
```diff
@@ -228,9 +228,12 @@ public function synthesize_speech( $post_id ) {
 		// Create the request body to synthesize speech from text.
 		$request_body = sprintf(
-			"<speak version='1.0' xml:lang='en-US'><voice xml:lang='en-US' xml:gender='%s' name='%s'>%s</voice></speak>",
+			apply_filters( 'classifai_azure_text_to_speech_ssml_tag', $ssml_tag ),
```
@dkotter Do you think we can keep this filter hook? It is not related to the pull request context.

dkotter replied:
I wonder if this was added to allow people to filter the `xml:lang` part? If not, I think this can be removed. If we do keep it though, we'll want a filter docblock added here

ravinderk replied:
I think we can remove it for now, and We should do it in a separate pull request.

ravinderk committed:
format: improve code formatting

ravinderk commented:
@dkotter This pull request is ready for review.

dkotter commented on includes/Classifai/Providers/OpenAI/ChatGPT.php:
```diff
@@ -1018,4 +1048,31 @@ public function get_content( int $post_id = 0, int $return_length = 0, bool $use
+		// Use English name for language if available.
+		if ( array_key_exists( $language_code, $translations ) ) {
+			$language_english_name = $translations[ $language_code ]['english_name'] ?? '';
+			$language              = " in $language_english_name  language";
```
Seems like extra space here:
```suggestion
			$language              = " in $language_english_name language";
```

dkotter commented on includes/Classifai/Providers/OpenAI/ChatGPT.php:
```diff
@@ -1018,4 +1048,31 @@ public function get_content( int $post_id = 0, int $return_length = 0, bool $use
+	 * @since x.x.x
+	 * @param string $language_code Language code.
+	 */
+	private function get_language_for_prompt( string $language_code ): string {
```
I almost think most of this should be moved into a shared utility function (so outside of ChatGPT and OpenAI). I say this because I can imagine other providers/features wanting to be able to get the full language name from a language code.

I would move that portion of this code into a shared helper function and than this method can then just call that utility function and return the `" in $language_english_name language"` part, if that makes sense

ravinderk committed:
Update includes/Classifai/Providers/OpenAI/ChatGPT.php

Co-authored-by: Darin Kotter <darin.kotter@gmail.com>

ravinderk committed:
refactor: remove unrelated changes

ravinderk committed:
refactor: create utility function

ravinderk commented:
@dkotter I implemented your suggestion. You can continue review.

jeffpaul commented:
Note that this work should be placed on hold, thus labeled as `BLOCKED`, as we work on #587 / #594 as this alternate approach allows for folks to tweak the language as part of editing the prompts for OpenAI.  We'll likely still want to assess adding in alternate language support for Watson and Azure tagging in posts and images, but I can leave a comment to that effect on the source Multilingual issue.

jeffpaul commented:
One thing I think might be worth testing before taking this work much further would be testing ClassifAI and changing the default prompt to be in a different language than the default English and see if OpenAI is "smart" enough to then return with content in that same language?

The non-OpenAI features would probably need an actual setting to tweak the input/output to something besides English, but I suspect we might get multilingual support with OpenAI much easier now that the default prompts are exposed and could be changed from English to non-English.

dkotter commented:
Yeah, this PR adds a language setting and then adjusts the prompt to tell ChatGPT to return responses in that language. But that can now be done by manually adjusting the prompt itself so I think this PR can be closed
